src/nss-resolve/nss-resolve.c

```diff
@@ -17,6 +17,15 @@
 #include "strv.h"
 #include "varlink.h"
 
+static JsonDispatchFlags json_dispatch_flags = 0;
+
+__attribute__((constructor)) static void setup_logging(void) {
```

Comment:
gaaaaaa, noooooo, not gcc constructors, please not. you never know when those get executed, an things are totally weird in threaded envs, and since this one depends on external conditions (notably env vars) that processes can set, I think we should just make this an idempotent function that is called at the beginning of each of the entrypoints we expose. It's not that it is like 500 entrypoints, it's just a handful, so it's easy to just call into one small helper function...

There should be some minimal consideration of multi-threaded behaviour here I figure. We never know in what context our NSS modules runs after all...

Reply:
Well, sure, we don't know when *exactly* they will be executed, but we don't care. Some time after dlopen() starts, some time before the function is executed. And yes, we don't know when exactly getenv() will be executed, but why would we even care? If you call setenv() concurrently from another thread, you'll get either the old or the new value. That's just fine.

Overall, I find this approach elegant. Is there any specific scenario where things go wrong with this approach?

Reply:
gcc constructors are horrible wherever used, I'd avoid them like the plague (remember libselinux). If we can solve it easily in a different way I'd always go the simpler way...

constructors are called from dlopen(), but it's not clear what actually happens if multiple threads do NSS and thus call dlopen(), in which order are they run, is there a mutex taken?

I'd be OK (though still no fan) if the gcc constructor would do trivial stuff, like initializing some vars or so, but that's not the case here actually. it prses /proc/cmdline and what not, i.e. does all kinds of hidden kernel API stuff at the moment of dlopen().

We ourselves manually load NSS modules at times to see what is available (nss-myhostname), it kinds sucks if that suddenly dlopen() means you access all kinds of stuff, instead of just checking the module and loading it into memory.

Code with hidden side-effects (i.e. beyond mere static initialization) is just bad I think. NSS modules are blackboxes that people have to trust. If we do hidden stuff that leaks outside of the process from the mere acting of a superficially innocent dlopen() then that sucks.

Really, no fan of gcc constructors. They aren't "elegant", they are a magic side-effect monster...

Thing is, the entrypoints we define already come with common prefix code, i.e. PROTECT_ERRNO() and BLOCK_SIGNALS(). Maybe we should have macro that merges these two plus the log initialization into one new macro: NSS_ENTRYPOINT_BEGIN() or so, to be defined in our nss-util.h, which not only removes the side-effect monster, but also saves code!

Reply:
Updated.